yt_dlp/networking/exceptions.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import annotations
+
+import http.client
+import urllib.error
+
+from ..utils import YoutubeDLError
+import typing
+if typing.TYPE_CHECKING:
+    from .response import Response
+    from . import RequestHandler
+    from typing import List, Union
+
+
+class RequestError(YoutubeDLError):
+    def __init__(
+        self,
+        msg: str = None,
+        cause: Union[Exception, str, None] = None,
+        handler: RequestHandler = None
+    ):
+        self.handler = handler
+        self.cause = cause
+        if not msg and cause:
+            msg = str(cause)
+        super().__init__(msg)
+
+
+class UnsupportedRequest(RequestError):
+    """raised when a handler cannot handle a request"""
+    pass
+
+
+class NoSupportingHandlers(RequestError):
+    """raised when no handlers can support a request for various reasons"""
+    def __init__(self, unsupported_errors: List[UnsupportedRequest], unexpected_errors: List[Exception]):
+        self.unsupported_errors = unsupported_errors or []
+        self.unexpected_errors = unexpected_errors or []
+
+        # Print a quick summary of the errors
+        err_handler_map = {}
+        for err in unsupported_errors:
+            err_handler_map.setdefault(err.msg, []).append(err.handler.RH_NAME)
+
+        reason_str = ', '.join([f'{msg} ({", ".join(handlers)})' for msg, handlers in err_handler_map.items()])
+        if unexpected_errors:
+            reason_str = ' + '.join(filter(None, [reason_str, f'{len(unexpected_errors)} unexpected error(s)']))
+
+        err_str = 'Unable to handle request'
+        if reason_str:
+            err_str += f': {reason_str}'
+
+        super().__init__(msg=err_str)
```

Comment:
This builds a generic message from multiple UnsupportedErrors, like:

`Unable to handle request: Unsupported url scheme "file" (urllib, requests) + 2 unexpected error(s)`

This is used in the director to raise an error when no handler is supported, using the reasons gathered. open to suggestions.

Reply:
I feel like this is overcomplicated, we can use a couple modifications to simplify this:
```py
    @classmethod
    def build(cls, errors: dict[Handler, Exception]):
        message = 'Cannot handle request'

        if errors:
            error_map = defaultdict(list)
            for handler, error in errors.items():
                message = error.msg if isinstance(error, UnsupportedRequest) else 'Unexpected error'
                error_map[message].append(handler.RH_NAME)

            reasons = ', '.join(f'{msg} ({", ".join(handlers)})' for msg, handlers in error_map.items())
            message = f'{message}: {reasons}'

        return cls(msg=message)
```
Using a `build` classmethod seems cleaner as it doesnt violate Liskov.
Which will produce the following output:

`Cannot handle request: Unsupported url scheme "file" (urllib, requests), Unexpected error (httpx)`

This would require a slight modification of the director to store a dict, and the classes (classes, not instances) to be hashable.

Reply:
that looks cleaner

> This would require a slight modification of the director to store a dict, and the classes (classes, not instances) to be hashable.

we could have the key be the RH_NAME to keep it simple.

I wanted to store the actual exceptions within this error so that extra processing can be done in YoutubeDL if needed.

Reply:
We need to store the original exceptions for further processing in YoutubeDL and such.

I'll leave this open but it's not a blocker for merge, can tidy it later